browser/registry.go

```diff
@@ -148,38 +154,136 @@ func (r *remoteRegistry) isRemoteBrowser() (string, bool) {
 	return wsURL, true
 }
 
-// browserRegistry stores browser instances indexed per
-// iteration as identified by VUID-scenario-iterationID.
+// browserRegistry stores a single VU browser instances
+// indexed per iteration.
 type browserRegistry struct {
+	vu k6modules.VU
+
 	mu sync.RWMutex
-	m  map[string]api.Browser
+	m  map[int64]api.Browser
+
+	buildFn browserBuildFunc
+}
+
+type browserBuildFunc func(ctx context.Context) (api.Browser, error)
+
+func newBrowserRegistry(vu k6modules.VU, remote *remoteRegistry, pids *pidRegistry) *browserRegistry {
+	bt := chromium.NewBrowserType(vu)
+	builder := func(ctx context.Context) (api.Browser, error) {
+		var (
+			err                    error
+			b                      api.Browser
+			wsURL, isRemoteBrowser = remote.isRemoteBrowser()
+		)
+
+		if isRemoteBrowser {
+			b, err = bt.Connect(ctx, wsURL)
+			if err != nil {
+				return nil, err //nolint:wrapcheck
+			}
+		} else {
+			var pid int
+			b, pid, err = bt.Launch(ctx)
+			if err != nil {
+				return nil, err //nolint:wrapcheck
+			}
+			pids.registerPid(pid)
+		}
+
+		return b, nil
+	}
+
+	r := &browserRegistry{
+		vu:      vu,
+		m:       make(map[int64]api.Browser),
+		buildFn: builder,
+	}
+
+	iterSubID, eventsCh := vu.Events().Local.Subscribe(
+		k6event.IterStart,
+		k6event.IterEnd,
+	)
+	unsubscribe := func() {
+		vu.Events().Local.Unsubscribe(iterSubID)
+	}
+	go r.handleIterEvents(eventsCh, unsubscribe)
+
+	return r
 }
 
-func newBrowserRegistry() *browserRegistry {
-	return &browserRegistry{
-		m: make(map[string]api.Browser),
+func (r *browserRegistry) handleIterEvents(eventsCh <-chan *k6event.Event, unsubscribeFn func()) {
+	var (
+		ok    bool
+		data  k6event.IterData
+		ctx   = context.Background()
+		vuCtx = k6ext.WithVU(r.vu.Context(), r.vu)
+	)
+
+	for e := range eventsCh {
+		// If browser module is imported in the test, NewModuleInstance will be called for
+		// every VU. Because on VU init stage we can not distinguish to which scenario it
+		// belongs or access its options (because state is nil), we have to always subscribe
+		// to each VU iter events, including VUs that do not make use of the browser in their
+		// iterations.
+		// Therefore, if we get an event that does not correspond to a browser iteration, then
+		// unsubscribe for the VU events and exit the loop in order to reduce unuseful overhead.
+		if !isBrowserIter(r.vu) {
+			unsubscribeFn()
+			e.Done()
+			return
+		}
+
+		if data, ok = e.Data.(k6event.IterData); !ok {
+			k6ext.Abort(vuCtx, "unexpected iteration event data format: %v", e.Data)
+			e.Done()
+			continue
+		}
+
+		switch e.Type { //nolint:exhaustive
```

Comment:
In this case the events are constrained by our own implementation when subscribing, so this error would not be much related to our implementation but with a possible k6-core error. Still it might make sense though. What do you think?

If we want to log there are various things to consider though:
We can not create an instance of our logger implementation in the browser registry constructor, as at that time the VU state is `nil` so we can not access the underlying k6 logger. That means that we either will have to:
- Create an instance of our logger for each received event. Which does not seem ideal.
- Use the VU logger such as `r.vu.State().Logger.Warnf("received unexpected event type: %T", e)`
- Use the std logger

Probably using the VU logger directly is the best option here, but that might imply that we miss some of the labels that are centralized in our logger wrapper implementation, although we can argue that this log is not strictly related with the browser behavior, but with a more "internal behavior/implementation".

What is your take on this?

Reply:
> so this error would not be much related to our implementation but with a possible k6-core error. Still it might make sense though. What do you think?

Yeah, agreed that it's not related to our implementation, but unexpected events could have a negative affect or at least help us debug an issue with the event system in the future if something changes. My preference would be to log the unexpected event here still.

> If we want to log there are various things to consider though

At the moment a `logger` instance is created when `Connect` or `Launch` are called, but could be nice to create a `logger` here and pass it on to those methods; it would require a tiny bit of refactoring though which possibly isn't suited for this PR, WDYT?

Otherwise I'm happy to go with `r.vu.State().Logger.Warnf("received unexpected event type: %v", e.Type)` for now.

Reply:
Yep, I don't think that refactor is small enough to include it in this PR. I think using the VU logger is the best option now.
Should the message be `received unexpected event type` or `received unsubscribed event type` as that probably is more specific and understandable to be in the context of the k6 event system?

Reply:
I'm leaning towards `received unexpected event type`. When i read `received unsubscribed event type`, i read it as the event system has emitted an `unsubscribed` event 😅

Reply:
Added. See [here](https://github.com/grafana/xk6-browser/pull/944/commits/a84847afe5637f7f425f569714e2c3cf269adb68#diff-f2c9c90a415d446c8ce22e2d961cf6c20fe8327cbf408087ed7abf1eeb5c256cR259-R261) in [a84847a](https://github.com/grafana/xk6-browser/pull/944/commits/a84847afe5637f7f425f569714e2c3cf269adb68).